Load Testing Using Locust

JosephSemrai opened the pull request:
## Description
Add load testing to various endpoints using the Locust Framework.
**The Google Code-in task has specified updating the wiki page. I am unable to update the wiki page, but you can find the relevant documentation pertaining to the addition of this feature at [Locust Wiki Page](https://github.com/JosephSemrai/Bassa-Load-Testing-Wiki/edit/master/README.md).

In order to successfully build the Docker container, the core requirements were updated.

## Related Issue
Google Code-in: [Bassa] Perform load testing using locust framework on production Bassa API container

## Motivation and Context
To determine how well the production Bassa API service handles traffic in terms of its Flask socket-io server.

The following API endpoints are tested:
- /login
- /api/user/requests
- /api/user/blocked
- /api/user/downloads
- /api/downloads
- /api/user/heavy

## How Has This Been Tested?
I had tested this code by running the full production container:
```bash
docker-compose up --build
```
and then executing the Locust tests in the **./bassa/components/core/tests** directory:
```bash
locust
```

## Screenshots (In case of UI changes):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

JosephSemrai commented:
For convenience, find the Bassa Wiki Page below at (as I am unable to make contributions to the wiki directly):
https://github.com/JosephSemrai/Bassa-Load-Testing-Wiki

kmehant commented on components/core/test-locust.sh:
```diff
@@ -0,0 +1,15 @@
+else
+    sudo apt-get install docker
+fi
+echo Building and running container
```
No need to install docker, We assume that the developer is ready with the setup

kmehant commented on components/core/test-locust.sh:
```diff
@@ -0,0 +1,15 @@
+echo Building and running container
+cd ../..
+echo Building from: $PWD...
+docker-compose -f docker-compose.yml up --build
```
Throw an error if this command fails, and report the developer that mostly has to install docker and docker-compose in the working machine pointing to some relevant docker installation guide.

JosephSemrai commented:
- Updated documentation images to reflect the production Docker build
- Added shell script to automate some of the testing process
- Added automated section to the documentation

kmehant commented on components/core/test-locust.sh:
```diff
@@ -0,0 +1,15 @@
+cd ../..
+echo Building from: $PWD...
+docker-compose -f docker-compose.yml up --build
+read -p "Done! Press [Enter] to continue..."
```
Here you can start locust

kmehant commented on components/core/test-locust.sh:
```diff
@@ -0,0 +1,15 @@
+#!/bin/bash
```
Shall we even move this to core/tests/

kmehant commented on components/core/test-locust.sh:
```diff
@@ -0,0 +1,15 @@
+    sudo apt-get install docker
+fi
+echo Building and running container
+cd ../..
```
I guess instead of changing the directory you can give the relative path to the docker-compose.yml file.

kmehant commented:
Finally, Some Codacy errors which you have to fix

JosephSemrai commented:
Hi,
I have gone through with the following:
- Adding error handlers for Locust and creating the Docker image
- Removing the install and OS check for docker
- Updating the paths to be relative (no need for cd anymore)
- Moving the test to the /tests directory
- Launching both Locust and docker-compose simultaneously requiring only one script

JaDogg commented on components/core/tests/locustfile.py:
```diff
@@ -0,0 +1,51 @@
+
+    @seq_task(1)
+    @task(1)
+    def userRequests(self):
```
can you change these functions to snake case 🥳 ex: `user_requests`

JaDogg replied:
Please do this for all your functions in this class.

kmehant commented on components/core/tests/test-locust.sh:
```diff
@@ -0,0 +1,23 @@
+    fi
+}
+
+start_docker(){
```
Please modify the function name to `start_bassa`

kmehant commented on components/core/tests/test-locust.sh:
```diff
@@ -0,0 +1,23 @@
+}
+
+start_locust &
+start_docker
```
You missed out this.

kmehant replied:
This will break the script change it to start_bassa

JosephSemrai replied:
Sorry about the delayed response. It was 4:00am locally, so I just committed the changes. The fix should be in the next commit.
